jetty-io/src/main/java/org/eclipse/jetty/io/BufferCallbackAccumulator.java

```diff
@@ -0,0 +1,93 @@
+//
+// ========================================================================
+// Copyright (c) 1995-2021 Mort Bay Consulting Pty Ltd and others.
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License v. 2.0 which is available at
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+// ========================================================================
+//
+
+package org.eclipse.jetty.io;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+
+import org.eclipse.jetty.util.BufferUtil;
+import org.eclipse.jetty.util.Callback;
+
+public class BufferCallbackAccumulator
+{
+    private final List<Entry> _entries = new ArrayList<>();
+
+    private static class Entry
+    {
+        private final ByteBuffer buffer;
+        private final Callback callback;
+
+        Entry(ByteBuffer buffer, Callback callback)
+        {
+            this.buffer = buffer;
+            this.callback = callback;
+        }
+    }
+
+    public void addEntry(ByteBuffer buffer, Callback callback)
+    {
+        _entries.add(new Entry(buffer, callback));
```

Comment:
Since we know that the length will ultimately be needed to take the content, we may as well keep a running total of the length here.  Then if it exceeds Integer.MAX_VALUE (or some other max), we can throw here rather than continue.

Reply:
I see the max size is protected in where this is used... but this class is in io utility package.